File: lt0108.cpp
#include <iostream>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x):val(x),left(nullptr), right(nullptr){}
};

class Solution {
public:

    using IterType = vector<int>::iterator;

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        
        if(nums.empty())
            return nullptr;

        int size = nums.size();
        int mid = size / 2;

        TreeNode* root = new TreeNode(nums[mid]);
        vector<int> left_tree = {nums.begin(), nums.begin() + mid};
        vector<int> right_tree = {nums.begin() + mid + 1, nums.end()};
        root->left = sortedArrayToBST(left_tree);
        root->right = sortedArrayToBST(right_tree);
        return root;
    }
};

int main()
{

}

File: code/lt0108.py
# https://leetcode-cn.com/problems/convert-sorted-array-to-binary-search-tree/ 
# convert sorted array to binary search tree 
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if len(nums) == 0:
            return None
        
        privot = len(nums) // 2

        root = TreeNode(nums[privot])

        root.left = self.sortedArrayToBST(nums[:privot])
        root.right = self.sortedArrayToBST(nums[privot + 1:])

        return root;